src/ssl.cpp

```diff
@@ -0,0 +1,257 @@
+are met:
+
+  * Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+  * Redistributions in binary form must reproduce the above copyright
+    notice, this list of conditions and the following disclaimer in
+    the documentation and/or other materials provided with the distribution.
+  * Neither the name of the author nor the names of its
+    contributors may be used to endorse or promote products derived
+    from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/ssl.hpp"
+
+#if TORRENT_USE_SSL
+
+#ifdef TORRENT_USE_OPENSSL
+#include <openssl/x509v3.h> // for GENERAL_NAME
+#endif
+
+#ifdef TORRENT_USE_GNUTLS
+#include <gnutls/x509.h>
+#endif
+
+namespace libtorrent {
+namespace ssl {
+
+void set_trust_certificate(context_handle_type c, string_view pem, error_code &ec)
+{
+#if defined TORRENT_USE_OPENSSL
+	// create a new X.509 certificate store
+	X509_STORE* cert_store = X509_STORE_new();
+	if (!cert_store)
+	{
+		ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+		return;
+	}
+
+	// wrap the PEM certificate in a BIO, for openssl to read
+	BIO* bp = BIO_new_mem_buf(
+			const_cast<char*>(pem.data()),
+			int(pem.size()));
+
+	// parse the certificate into OpenSSL's internal representation
+	X509* cert = PEM_read_bio_X509_AUX(bp, nullptr, nullptr, nullptr);
+	BIO_free(bp);
+
+	if (!cert)
+	{
+		X509_STORE_free(cert_store);
+		ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+		return;
+	}
+
+	// add cert to cert_store
+	X509_STORE_add_cert(cert_store, cert);
+	X509_free(cert);
+
+	// and lastly, replace the default cert store with ours
+	SSL_CTX_set_cert_store(c, cert_store);
+
+#elif defined TORRENT_USE_GNUTLS
+	std::string str(pem);
+    gnutls_datum_t ca;
+    ca.data = reinterpret_cast<unsigned char*>(const_cast<char*>(str.c_str())); // must be null terminated
+    ca.size = unsigned(str.size());
+
+	// Warning: returns the number of certificates processed or a negative error code on error
+	int ret = gnutls_certificate_set_x509_trust_mem(c.native_handle(), &ca, GNUTLS_X509_FMT_PEM);
```

Comment:
I thought it looked a bit weird that the `data` field needed to be null terminated, given that you pass in the length as well. So, I guess [this](https://www.gnutls.org/manual/html_node/Core-TLS-API.html#index-gnutls_005fcertificate_005fset_005fx509_005ftrust_005fmem) is the authoritative documentation.

I don't see it say anything about it needing to be 0-terminated. But I see `ca` is a *list* of certificates. How does it know how many certs are in the list? You pass in 1, but what stops the function from assuming there are more after? Are you sure it's not the list of certs that's null terminated? Say, the last `data` field is `nullptr` to indicate the end of the list.

If that is the case, you wouldn't have to copy the certificate above.

Reply:
Oh, that's true. Other gnutls x509 functions require it though, like [gnutls_certificate_set_x509_key_mem](https://www.gnutls.org/manual/html_node/Core-TLS-API.html#index-gnutls_005fcertificate_005fset_005fx509_005fkey_005fmem)

I'm pretty sure the fact that it can be a list in the case of PEM encoding only means you can list different CA in the same string.

Reply:
I see. Right, I think I was in C++ mode thinking that there must be a reason it was a pointer instead of being passed by value or reference :)